README.md
#chessbot

Modified from the [react tic-tac-toe tutorial](https://reactjs.org/tutorial/tutorial.html).

Interesting code to look at:

front-end/src/App.js getSlidingMoves function (line 171) specifically I think the double for loop at the end is particularly elegant.


back-end/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-03-30
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-03-30
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/rest-api/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/rest-api/compare/0.1.0...0.1.1


old/js/react_test.js
'use strict';

const e = React.createElement;

class LikeButton extends React.Component {
  constructor(props) {
    super(props);
    this.state = { liked: false };
  }

  render() {
    if (this.state.liked) {
      return 'You liked this.';
    }

    return e(
      'button',
      { onClick: () => this.setState({ liked: true }) },
      'Like'
    );
  }
}


const domContainer = document.querySelector('#boardtest');
ReactDOM.render(e(LikeButton), domContainer);


back-end/test/rest_api/core_test.clj
(ns rest-api.core-test
  (:require [clojure.test :refer :all]
            [rest-api.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


old/js/move.js
var selectedPiece = '';
var selectedId = '';
function getRookMoves(id){

}
function getBishopMoves(id){

}
function getKingMoves(id){

}
function getQueenMoves(id){
        return getRookMoves(id).concat(getBishopMoves(id));
}
function getKnightMoves(id){

}
function getPawnMoves(id){

}
function getLegalMoves(id, type){
        //make all legal moves
        return Array.from({length: 64}, (item, index) => index);
        if (Set(['r', 'R']).has(type)){
                return getRookMoves(id)
        }else if (Set(['b', 'B']).has(type)){
                return getBishopMoves(id)
        }else if (Set(['n', 'N']).has(type)){
                return getKnightMoves(id)
        }else if (Set(['q', 'Q']).has(type)){
                return getQueenMoves(id)
        }else if (Set(['k', 'K']).has(type)){
                return getKingMoves(id)
        }else if (Set(['p', 'P']).has(type)){
                return getPawnMoves(id)
        }else{
                alert("Something has gone wrong with get legal moves function")
        }




}
function getAllLegalMoves(){
        var map = [];
        for (var id = 0; id < 64; id++){
                map.push(getLegalMoves(id))
        }
        return map
}
function highlightLegalMoves(squares){
        var highlight = function (id) {document.getElementById(id).classList.add("highlighted")};
        squares.forEach(highlight);
}
function unhighlightEverything(){
        var unhighlight = function (id) {document.getElementById(id).classList.remove("highlighted")};
        Array.from({length: 64}, (item, index) => index).forEach(unhighlight);
}
function pickPiece(img){
        var id = img.parentNode.id;
        var type = img.id.charAt((img.id.length)-1);
        highlightLegalMoves(getLegalMoves(id, type));
        selectedPiece = type;
	selectedId = id;
}
function placePiece(td){
	console.log("placing piece");
	var id = td.id;
	console.log(td);
        if (selectedPiece != ''){
                console.log("unhighlighting");
                unhighlightEverything();
		document.getElementById(id).innerHTML = '<img class = "piece" id = "' + id + selectedPiece + '" onclick = "pickPiece(this)" src = img/' + selectedPiece + '.svg>';
        	selectedPiece = '';
	}
}



back-end/project.clj
(defproject rest-api "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 ;compojure - A basic routing library
                 [compojure "1.6.2"]
                 ;HTTP library for client/server
                 [http-kit "2.3.0"]
                 ;Ring defaults - for query params etc
                 [ring/ring-core "1.9.2"]
                 [ring/ring-jetty-adapter "1.9.2"]
                 ;necessary for ring
                 [javax.servlet/servlet-api "2.5"]
                 ;clojure data.json library
                 [org.clojure/data.json "2.0.2"]]
  :main ^:skip-aot rest-api.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}})


front-end/src/App.js
import './App.css';
import React from 'react';
import r_b from "./img/r.svg";
import n_b from "./img/n.svg";
import b_b from "./img/b.svg";
import q_b from "./img/q.svg";
import k_b from "./img/k.svg";
import p_b from "./img/p.svg";
import r_w from "./img/R.svg";
import n_w from "./img/N.svg";
import b_w from "./img/B.svg";
import q_w from "./img/Q.svg";
import k_w from "./img/K.svg";
import p_w from "./img/P.svg";

function legalMovesCheckSide(currentSide, squareSide){
	let add;
	let stop;
	if (squareSide === 'empty'){
		add = true;
		stop = false;
	}else if(squareSide === currentSide){
		add = false;
		stop = true;
	}else{
		add = true;
		stop = true;
	}
	return [add, stop]
}
function getNumSquaresToEdge(){
	var numSquaresToEdge = new Array(64);
	for (var file = 0; file < 8; file ++){
		for (var rank = 0; rank < 8; rank ++){
			var numNorth = rank;
			var numSouth = 7-rank;
			var numWest = file;
			var numEast = 7-file;
			var squareIndex = rank*8+file
			numSquaresToEdge[squareIndex] = new Map();
			numSquaresToEdge[squareIndex].set(-8, numNorth);
			numSquaresToEdge[squareIndex].set(8, numSouth);
			numSquaresToEdge[squareIndex].set(-1, numWest);
			numSquaresToEdge[squareIndex].set(1, numEast);
			numSquaresToEdge[squareIndex].set(-9, Math.min(numNorth, numWest));
			numSquaresToEdge[squareIndex].set(-7, Math.min(numNorth, numEast));
			numSquaresToEdge[squareIndex].set(9, Math.min(numSouth, numEast));
			numSquaresToEdge[squareIndex].set(7, Math.min(numSouth, numWest));
		}
	}
	return numSquaresToEdge
}
function Square(props){
	var classes = "square";
	((Math.floor(props.id/8) + (props.id % 8)) % 2 !== 0) ? classes += " black" : classes += " white"; 
	(props.isHighlighted) ? classes += " highlighted" : classes += "";
	let imgMap = new Map();
	imgMap.set('r', r_b);
	imgMap.set('n', n_b);
	imgMap.set('b', b_b);
	imgMap.set('q', q_b);
	imgMap.set('k', k_b);
	imgMap.set('p', p_b);
	imgMap.set('R', r_w);
	imgMap.set('N', n_w);
	imgMap.set('B', b_w);
	imgMap.set('Q', q_w);
	imgMap.set('K', k_w);
	imgMap.set('P', p_w);
	let imgsrc;
	(props.piece !== '') ? imgsrc = imgMap.get(props.piece) : imgsrc = '';
	//imgsrc = require("./logo.svg");
	return (
		<button className={classes} onClick={props.onClick}>
			<img className="piece" src={imgsrc} alt={props.piece}></img>
			{props.id}
		</button>
	);
}
function getSideFromPiece(piece){
	if (new Set(['r', 'n', 'b', 'q', 'k', 'p']).has(piece)){
		return 'black'
	}else if (new Set(['R', 'N', 'B', 'Q', 'K', 'P']).has(piece)){
		return 'white'
	}else{
		return 'empty'
	}
}
function getBoardFromFEN(FEN){
        var map = [];
        for (var i = 0; i<64; i++){
                map.push('');
        }
        var id = 0;
        const pieces = new Set(['r', 'n', 'b', 'q', 'k', 'p', 'R', 'N', 'B', 'Q', 'K', 'P']);
        var i = 0;
        while (FEN.charAt(i) !== ' '){
		var ch = FEN.charAt(i)
                if (pieces.has(ch)){
                        map[id] = ch;
                        id += 1;
                }else if(ch >= '0'.charAt(0) && ch <= '9'.charAt(0)){
                        id += Number(ch);
                }
                i++;
        }
        return map;
}
function calculateWinner(squares){
	return null;
	//can also return "draw"
}
function range(start, end) {
  return Array(end - start + 1).fill().map((_, idx) => start + idx)
}
class Board extends React.Component{
	constructor(props){
		super(props);
		this.state = {
			boardSize: 8,
		};
	}
	renderSquare(i) {
		return (<Square 
			isHighlighted={this.props.highlighted.has(i)}
			id={i}
			piece={this.props.squares[i]}
			onClick={() => this.props.onClick(i)}
		/>);
	}
	renderRow(rowNumber){
		const colNumbers = Array.from(Array(this.state.boardSize).keys());
		const squares = colNumbers.map((colNumber) =>
			this.renderSquare(rowNumber*this.state.boardSize + colNumber)
		);


		return(<div className="board-row">{squares}</div>);

	}
	render(){
		const numbers = Array.from(Array(this.state.boardSize).keys());
		const rows = numbers.map((number) =>
			this.renderRow(number)
		);
		return(
			<div>{rows}</div>
		)	
	}

}

class Game extends React.Component{
	constructor(props) {
		super(props);
		this.state = {
			history: [{
				FEN: "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
				squares: getBoardFromFEN("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"),
				whiteShortCastle: true,
				whiteLongCastle: true,
				blackShortCastle: true,
				blackLongCastle: true,
				enPassantTarget: null,
			}],
			whiteTurn: true,
			stepNumber: 0,
			selectedPiece: null,
			selectedID: null,
			numSquaresToEdge: getNumSquaresToEdge(),
		};
	}
	getSlidingMoves(id){
		const vertAndHor = [8, -8, 1, -1];
		const diagonals = [7,9,-7,-9];
		const squares = this.state.history[this.state.stepNumber].squares;
		const boundaries = new Set([0,1,2,3,4,5,6,7,8,15,23,31,39,47,55,63,62,61,60,59,58,57,56,48,40,32,24,16,8])
		const friendlySide = getSideFromPiece(squares[id]);
		var legalMoves = new Set();
		var dirs = [];
		var pointer = id
		let add;
		let stop;
		if (new Set(['b', 'B', 'q', 'Q']).has(squares[id])){
			diagonals.forEach(dir => dirs.push(dir))
		}
		if (new Set(['r', 'R', 'q', 'Q']).has(squares[id])){
			vertAndHor.forEach(dir => dirs.push(dir))
		}
		for (let dir of dirs){
			//var dir = dirs[d];
			pointer = id+dir;
			for (var i = 0; i < this.state.numSquaresToEdge[id].get(dir); i++){
				[add, stop] = legalMovesCheckSide(friendlySide, getSideFromPiece(squares[pointer]));	
				if (add) legalMoves.add(pointer);
				if (stop) break;
				pointer += dir;
			}
		}
		legalMoves.delete(id);
		return legalMoves
	}
	
	getKnightMoves(id){
		const paths = [[-1,-2,-10], [-1,-2,6], [-8,-16,-17], [-8,-16,-15], [1,2,-6], [1,2,10], [8,16,17], [8,16,15]]
		const squares = this.state.history[this.state.stepNumber].squares;
		const boundaries = new Set([0,1,2,3,4,5,6,7,8,15,23,31,39,47,55,63,62,61,60,59,58,57,56,48,40,32,24,16,8])
		const friendlySide = getSideFromPiece(squares[id]);
		var legalMoves = new Set();
		for (let path of paths){
			var [allowed,] = legalMovesCheckSide(friendlySide, getSideFromPiece(squares[id+path[2]]));
			var lastPointer = id;
			for (let position of path){
				var pointer = id+position;
				var lastRow = Math.floor(lastPointer/8);
				var lastFile = lastPointer % 8;
				var currentRow = Math.floor(pointer/8);
				var currentFile = pointer % 8;
				if (Math.abs(currentRow-lastRow) > 1) allowed = false
				if (Math.abs(currentFile-lastFile) > 1) allowed = false
				if (Math.abs(currentRow-lastRow) >= 1 && Math.abs(currentFile-lastFile) >= 1) allowed = false
				lastPointer = pointer;
			}
			if (allowed) legalMoves.add(id+path[2]);
		}
		return legalMoves
	}
	getKingMoves(id){
		const positions = [8, -8, 1, -1, 7, 9, -7, -9];
		const squares = this.state.history[this.state.stepNumber].squares;
		const friendlySide = getSideFromPiece(squares[id]);
		var legalMoves = new Set();
		let allowed;
		let pointer;
		for (var d = 0; d < positions.length; d++){
			pointer = id+positions[d];
			[allowed, ] = legalMovesCheckSide(friendlySide, getSideFromPiece(squares[pointer]))
			if (this.state.numSquaresToEdge[id].get(positions[d]) == 0) allowed = false;
			if (allowed) legalMoves.add(pointer)
		}
		return legalMoves
	}
	getPawnMoves(id){
		const squares = this.state.history[this.state.stepNumber].squares;
		const friendlySide = getSideFromPiece(squares[id]);
		let direction;
		(friendlySide === 'white') ? direction = 1 : direction = -1;
		var positions = new Array();
		//add 1 move forward if not blocked
		if (getSideFromPiece(squares[id-(8*direction)]) === 'empty') positions.push(-8)
		//add 2 move foward if on starting row
		if (direction === 1 && new Set(range(48, 55)).has(id)){
			positions.push(-16);
		}else if (direction === -1 && new Set(range(8,15)).has(id)){
			positions.push(-16);
		}
		//check if it can capture on left or right
		const rightSide = getSideFromPiece(squares[id+(-7*direction)])
		const leftSide = getSideFromPiece(squares[id+(-9*direction)])
		if (rightSide !== friendlySide && rightSide !== 'empty'){
			positions.push(-7);
		}
		if (leftSide !== friendlySide && leftSide !== 'empty'){
			positions.push(-9);
		}
		const legalMoves = new Set(positions.map((position) => id+(position*direction)));
		return legalMoves
	}
	getLegalMoves(id, piece){
		if (id !== null && piece !== ''){
			if (new Set(['r', 'R', 'b', 'B', 'q', 'Q']).has(piece)){
				return this.getSlidingMoves(id)
			}else if (new Set(['n', 'N']).has(piece)){
				return this.getKnightMoves(id)
			}else if (new Set(['k', 'K']).has(piece)){
				return this.getKingMoves(id)
			}else if (new Set(['p', 'P']).has(piece)){
				return this.getPawnMoves(id)
			}else{
				console.log("Something has gone wrong with get legal moves function")
				return new Set();
			}

		}else{
			return new Set()
		}
	}
	makeRandomMove(side){
		const history = this.state.history.slice(0, this.state.stepNumber+1);
		const current = history[history.length-1];
		var squares = current.squares.slice();
		var possibleMoves = new Array(64);
		if (side == 'black'){
			possiblePieces = new Set(['r', 'b', 'q', 'n', 'k', 'p'])
		}
		else if (side == 'white'){
			possiblePieces = new Set(['R', 'B', 'Q', 'N', 'K', 'P'])
		}
		else{
			throw new Error("Make random move was given a side that was not black or white");
		}
		for (var i = 0; i < 64; i++){
			possibleMoves.append(getLegalMoves(i, squares[i]));
		}
		console.log(possibleMoves);
	}
	handleClick(i){
		const history = this.state.history.slice(0, this.state.stepNumber+1);
		const current = history[history.length-1];
		var squares = current.squares.slice();
		const selectedPiece = this.state.selectedPiece;
		const selectedID = this.state.selectedID;
		if (calculateWinner(squares)) {
			return;
		}
		//if already selected a piece
		else if(selectedPiece && selectedID != null){
			//if clicking on selected piece deselect it
			if (selectedID == i){
				this.setState({
					selectedPiece: null,
					selectedID: null,
				});
			}
			//if i is a legal move
			else if (this.getLegalMoves(selectedID, selectedPiece).has(i)){
				squares[selectedID] = "";
				squares[i] = selectedPiece;
				this.setState({
					history: history.concat([{
						squares: squares,
					}]),
					whiteTurn: !this.state.whiteTurn,
					stepNumber: history.length,
					selectedPiece: null,
					selectedID: null,
				});
			
			}

		}
		else if (this.getLegalMoves(i, squares[i]).size !== 0 && (this.state.whiteTurn === (getSideFromPiece(squares[i]) === 'white'))){
			this.setState({
				selectedPiece: squares[i],
				selectedID: i,
			});
		}
		
	}
	jumpTo(step){
		this.setState({
			stepNumber: step,
			whiteTurn: (step % 2) === 0,
		});
	}
	render(){
		const history = this.state.history;
		const current = history[this.state.stepNumber];
		const winner = calculateWinner(current.squares);
		const moves = history.map((step, move) => {
			const desc = move ?
				'Go to move #' + move:
				'Go to game start';
			if (move === this.state.stepNumber){
				return (
					<li key = {move}>
						<button onClick={() => this.jumpTo(move)}><b>{desc}</b></button>
					</li>
				);
			}else{
				return(
					<li key = {move}>
						<button onClick={() => this.jumpTo(move)}>{desc}</button>
					</li>

				);
			}
		});
		let status;
		if (winner === "draw"){
			status = "Draw";
		}
		else if (winner){
			status = 'Winner: ' + winner;
		}
		else {
			status = (this.state.whiteTurn ? 'White': 'Black') + "'s turn";
		}
		return(
			<div className="game">
				<div className = "game-board">
				<Board 
					squares={current.squares}
					highlighted={this.getLegalMoves(this.state.selectedID, this.state.selectedPiece)}
					onClick={(i) => this.handleClick(i)}
				/>
				</div>
				<div className = "game-info">
					<div>{status}</div>
					<ol>
						{moves}
					</ol>
				</div>
			</div>
		);
	}
}

function App() {
  return (
  	<Game />
  );
}

export default App;


back-end/src/rest_api/core.clj
(ns rest-api.core
  (:require [org.httpkit.server :as server]
            [compojure.core :refer :all]
            [compojure.route :as route]
            [ring.core.protocols :refer :all]
            [ring.adapter.jetty :refer :all]
            [ring.middleware.params :refer :all]
            [clojure.string :as str]
            [clojure.pprint :as pp]
            [clojure.data.json :as json])
  (:gen-class))
(def boundaries (set '(0,1,2,3,4,5,6,7,8,15,23,31,39,47,55,63,62,61,60,59,58,57,56,48,40,32,24,16,8)))
(def lengthToEdge
  (to-array 
    (for [rank (range 8) file (range 8)]
      (let [north rank
            south (- 7 rank)
            west file
            east (- 7 file)]
        (hash-map
            -8 north
            8 south
            -1 west
            1 east
            -9 (min north west)
            -7 (min north east)
            9 (min south east)
            7 (min south west)
        )
      )
    )
  )
)
(defn getKingMoves [id squares castling]
  (let [checkCastelClear (fn [start end] (every? (fn [pointer] (= "" (get square pointer))) (range (+ 1 start) end)))
        ;order is whiteLong, whiteShort, blackLong, blackShort
        clearChecks '[(checkCastelClear 56 60) (checkCastelClear 60 63) (checkCastelClear 0 4) (checkCastelClear 4 7)]
        castlingCheck '[(contains? castling \Q) (contains? castling \K) (contains? castling \q) (contains? castling \k)]
        canCastle (mapv (fn [one two] (and one two)) clearChecks castlingCheck)
        castleAnswers (cond
                   (= 1 (getSide (get squares id)))
                    (cond 
                      (and (get canCastle 0) (get canCastle 1)) '[62 58]
                      (get canCastle 0) '[58]
                      (get canCastle 1) '[62]
                      :else '[])
                   (= -1 (getSide (get squares id)))
                    (cond
                      (and (get canCastle 2) (get canCastle 3)) '[2 6]
                      (get canCastle 2) '[2]
                      (get canCastle 3) '[6]
                      :else '[]) 
                   :else '[])
        dirs '[-8 8 -1 1 -9 9 -7 7]
        side (getSide (get squares id))]
    (loop [i_dirs 0
           answers castleAnswers]
      (cond
        (= i_dirs (- (count dirs) 1)) answers
        (not= side (getSide squares (+ id (get dirs i_dirs)))) (recur (inc i_dirs) (conj answers (+ id (get dirs i_dirs))))
        :else (recur (inc i_dirs) answers)
        )
      )
    )    
  )
(defn getSide [square]
  (cond
    (contains? (set '(\r \n \b \q \k \p)) square) -1
    (contains? (set '(\R \N \B \Q \K \P)) square) 1
    :default 0
  )
)


(defn getSlidingMoves [id squares]
  
  (let [dirs (cond 
               (contains? (set '(\Q \q)) (get squares id)) '[8 -8 1 -1 7 -7 9 -9]
               (contains? (set '(\B \b)) (get squares id)) '[7 -7 9 -9]
               (contains? (set '(\R \r)) (get squares id)) '[8 -8 1 -1]
               :else (throw (Exception. (str "Called get sliding moves for non-sliding piece with id: " id)))
              )
        checkDir (fn [id dir squares side]
                (loop [answers '[]
                       distance 1]
                      (let [pointer (+ id (* distance dir))
                            pointerSide (getSide (get squares pointer))]
                        (cond
                          (= side pointerSide) answers
                          (= (* -1 side) pointer side) (conj answers pointer)
                          (<= (get (get lengthToEdge id) dir) distance) (conj answers pointer)
                          :else (recur (conj answers pointer) (+ distance 1))))))
            ]
    (reduce concat (for [dir dirs] (checkDir id dir squares (getSide (get squares id)))))
    )
)
(defn getPawnMoves [id squares enPassant]
  )
(defn getKnightMoves [id squares]
  )
(defn checkLegality [id squares]
  (cond
    ;(contains? (set '("K" "k"))) true
    
    )
  )
(defn truthy? [s]
  (cond
    (contains? (set '("false" "False" "f" "F")) s) false
    (contains? (set '("true" "True" "t" "T")) s) true
    :default (throw (Exception. (str "Truthy? could not evaluate the string: " s)))))

(defn FENtoSquares [FEN & {:keys [squares] :or {squares '[]}}]
  (loop [FEN FEN
         squares squares]
    (cond
      (or (= \space (get FEN 0)) (= 0 (count FEN))) squares
      (= \/ (get FEN 0)) (recur (subs FEN 1 (count FEN)) squares)
      (not= nil (re-find #"[0-9]" (str (get FEN 0)))) (recur (subs FEN 1 (count FEN)) (vec (concat squares (repeat (- (int (get FEN 0)) 48) ""))))
      :default (recur (subs FEN 1 (count FEN)) (conj squares (get FEN 0)))
      )
    )
  )
(defn simple-body-page [req]
  {:status  200
   :headers {"Content-Type" "text/html"}
   :body    "Hello World"})

(defn move-request [req]
  (println "request: " req)
  (println "HEADERS: " (req :headers))
  (let [params ((assoc-query-params req (or (req :character-encoding) "UTF-8")) :params)
        FEN (params "FEN")
        squares (FENtoSquares FEN)
        activeSide (get (str/split FEN #" ") 1)
        castling (get (str/split FEN #" ") 2)
        enPassant (get (str/split FEN #" ") 3)
        halfMove (get (str/split FEN #" ") 4)
        fullMove (get (str/split FEN #" ") 5)
        ]
    (println squares)
    (println (get lengthToEdge 10))
    (println "Sliding moves: " (getSlidingMoves 10 squares))
    {:status  200
     :headers {"Content-Type" "application/json"}
     :body    (json/write-str {:hello "HELLO"})}))


(defroutes app-routes
    (GET "/move-request" [] move-request)
    (route/not-found "Error, page not found!")
)

(defn -main
  [& args]
  (let [port (Integer/parseInt (or (System/getenv "PORT") "3001"))]
   (run-jetty move-request {:port port
                             :join? false})
    (println (str "Running webserver at http://127.0.0.1:" port "/"))))


old/js/init.js
var table = document.createElement("table");
table.id = 'board'
const FEN_Start = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"
var i_FEN = 0
const pieces = new Set(['r', 'n', 'b', 'q', 'k', 'p', 'R', 'N', 'B', 'Q', 'K', 'P']);
function getBoardFromFEN(FEN){
	var map = [];
	for (var i = 0; i<64; i++){
		map.push('');
	}
	var id = 56;
	const pieces = new Set(['r', 'n', 'b', 'q', 'k', 'p', 'R', 'N', 'B', 'Q', 'K', 'P']);
	var i = 0;
	while (FEN.charAt(i) != ' '){
		var ch = FEN.charAt(i)
		if (pieces.has(ch)){
			map[id] = ch;
			id += 1
		}else if(ch == '/'.charAt(0)){
			id -= 16;
		}else if(ch >= '0'.charAt(0) && ch <= '9'.charAt(0)){
			id += Number(ch);
		}
		i++;
	}
	return map;
}
 
for (var i = 0; i < 8; i++) {
	var tr = document.createElement('tr');
	for (var j = 0; j < 8; j++) {
		var td = document.createElement('td');
		if (i%2 == j%2) {
			td.className = "white square";
			td.id = 8*(7-i) + j;

		} else {
			td.className = "black square";
			td.id = 8*(7-i) + j;
		}
		td.addEventListener("click", placePiece);
		td.innerHTML = td.id
		tr.appendChild(td);
	}
	table.appendChild(tr);
}
document.body.appendChild(table);
pieceMap = getBoardFromFEN(FEN_Start);
for (var i = 0; i < 64; i++){
	if (pieces.has(pieceMap[i])) document.getElementById(i).innerHTML = '<img class = "piece" id = "' + i + pieceMap[i] + '" onclick = "pickPiece(this)" src = img/' + pieceMap[i] + '.svg>';

}

